### ISSUE TITLE
Implement suggestions from cypress retry-ability docs

### ISSUE THREAD

**Issue by ismay** [dhis2/ui#103]
From here: https://docs.cypress.io/guides/core-concepts/retry-ability.html.

Cypress will retry  commands that are followed by assertions until the assertion passes (if they query the DOM). So it's good practice to follow any Cypress commands that query the DOM with assertions, so that our tests don't trip over any async behaviour in our components.

So for example:

```js
it('creates 2 items', () => {
  cy.visit('/')                       // command
  cy.focused()                        // command
    .should('have.class', 'new-todo') // assertion

  cy.get('.new-todo')                 // command
    .type('todo A{enter}')            // command
    .type('todo B{enter}')            // command

  cy.get('.todo-list li')             // command
    .should('have.length', 2)         // assertion
})
```

See the docs for more details: https://docs.cypress.io/guides/core-concepts/retry-ability.html#Built-in-assertions

**Comment by Mohammer5** [comment]
> So it's good practice to follow any Cypress commands that query the DOM with assertions

This is not accurate. There are commands that have a built-in retry-ability:

> Often a Cypress command has built-in assertions that will cause the command to be retried.

Regarding what commands have a built-in retry ability, they say this:

> Cypress only retries commands that query the DOM: `cy.get()`, `.find()`, `.contains()`, etc. You can check if a particular command is retried by looking at the “Assertions” section in its API documentation.

Your suggestion is to add an assertion after `any Cypress commands that queries the DOM`. But the documentation clearly states that all commands that query the DOM already retry until the timeout is reached. It's better to know the commands we're using and use an assertion only when necessary instead of all the time.

**Comment by HendrikThePendric** [comment]
This will make it into the 5.0 milestone, see https://github.com/dhis2/ui/pull/102

**Comment by ismay** [comment]
> > So it's good practice to follow any Cypress commands that query the DOM with assertions
> 
> This is not accurate. There are commands that have a built-in retry-ability:
> 
> > Often a Cypress command has built-in assertions that will cause the command to be retried.
> 
> Regarding what commands have a built-in retry ability, they say this:
> 
> > Cypress only retries commands that query the DOM: `cy.get()`, `.find()`, `.contains()`, etc. You can check if a particular command is retried by looking at the “Assertions” section in its API documentation.
> 
> Your suggestion is to add an assertion after `any Cypress commands that queries the DOM`. But the documentation clearly states that all commands that query the DOM already retry until the timeout is reached. It's better to know the commands we're using and use an assertion only when necessary instead of all the time.

I think we're misunderstanding eachother. What I meant is basically what they're saying here in the cypress docs:

<img width="654" alt="Screenshot 2020-05-25 at 14 05 17" src="https://user-images.githubusercontent.com/7355199/82811352-db4da700-9e90-11ea-95b6-259d773ce306.png">

You're right in saying that the commands that query the DOM will be retried by default. But say you're expecting two DOM nodes, but `.get` only finds one, if there's no assertion following it cypress will assume the command has succeeded and move on.

To clarify, my suggestion with this issue is basically to follow the recommendations from: https://docs.cypress.io/guides/core-concepts/retry-ability.html#Timeouts (read from there). There they state that you should either create a single chain (so that the entire chain will be retried), or follow DOM querying commands with assertions (where needed), to ensure everything behaves as expected.

**Comment by ismay** [comment]
I've changed the title a bit to clarify my intention a bit more clearly (hopefully). So to summarize, suggestions from the article to avoid flaky tests:

1. Avoid unnecessarily splitting commands that query elements: https://docs.cypress.io/guides/core-concepts/retry-ability.html#Merging-queries
2. Alternate commands with assertions: https://docs.cypress.io/guides/core-concepts/retry-ability.html#Alternate-commands-and-assertions

**Comment by HendrikThePendric** [comment]
#87 was closed by #102 and this represented the move from `.then()` to `.should()`, which was actually causing our tests to produce inconsistent outcomes, and therefore had to be implemented for `v5.0.0`. The remaining work in this issue can be done for `v5.0.1` so I'm assigning this issue to that.

**Comment by varl** [comment]
https://jira.dhis2.org/browse/LIBS-61
